[Biomedical] Test Case Randomly Failing

I've seen ```testUnitParabolaString2D``` randomly fail in the CI with the same message a couple of times now. It failed on Windows most recently, but it might not be limited to that platform. Here's the traceback from the CI:
```
 FAIL: testUnitParabolaString2D (apply_parabolic_inlet_process_test.ApplyParabolicInletProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\Kratos\Kratos\bin\Custom\applications\FluidDynamicsBiomedicalApplication\tests\apply_parabolic_inlet_process_test.py", line 180, in testUnitParabolaString2D
    self.assertVectorAlmostEqual(test_node.GetSolutionStepValue(KratosMultiphysics.VELOCITY), [0.0,2.0,0.0])
  File "D:\a\Kratos\Kratos\bin\Custom\KratosMultiphysics\KratosUnittest.py", line 92, in assertVectorAlmostEqual
    self.assertAlmostEqual(v1, v2, places, LazyErrMsg(i, msg), delta)
AssertionError: 0.0 != 2.0 within 7 places (2.0 difference) : 
Check failed because vector arguments are not equal in component 1
Vector 1:
[3](-0,0,-0)
Vector 2:
[0.0, 2.0, 0.0]
```

@rubenzorrilla can you please look into it?

--- a/applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.cpp
+++ b/applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.cpp
@@ -205,7 +205,7 @@ void ParabolicProfileUtilities::ImposeParabolicInlet(
 template<class TInputType>
 void ParabolicProfileUtilities::ImposeParabolicProfile(
     ModelPart &rModelPart,
-    const TInputType & rMaxParabolaValue,
+    const TInputType& rMaxParabolaValue,
     const double MaxValueFactor)
 {
     // Get time value from model part ProcessInfo

--- a/applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.cpp
+++ b/applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.cpp
@@ -199,15 +199,18 @@ void ParabolicProfileUtilities::ImposeParabolicInlet(
     const double MaxValueFactor)
 {
     // Impose the parabolic inlet profile
-    ImposeParabolicProfile(rModelPart, MaxParabolaValue, MaxValueFactor);
+    ImposeParabolicProfile(rModelPart, *MaxParabolaValue, MaxValueFactor);
 }
 
 template<class TInputType>
 void ParabolicProfileUtilities::ImposeParabolicProfile(
     ModelPart &rModelPart,
-    const TInputType& rMaxParabolaValue,
+    const TInputType & rMaxParabolaValue,
     const double MaxValueFactor)
 {
+    // STOP
+    std::cout << "\nPROCESS BEGIN\n" << std::endl;
+
     // Get time value from model part ProcessInfo
     const auto& r_process_info = rModelPart.GetProcessInfo();
     KRATOS_ERROR_IF_NOT(r_process_info.Has(TIME)) << "TIME is not present in '" << rModelPart.FullName() << "' ProcessInfo." << std::endl;
@@ -219,28 +222,38 @@ void ParabolicProfileUtilities::ImposeParabolicProfile(
     });
     KRATOS_ERROR_IF(std::abs(max_dist) < 1.0e-12) << "WALL_DISTANCE is close to zero." << std::endl;
 
+    // TLS with the Inlet normal and the rMaxParabolaValue (to prevent race conditions if its a function)
+    struct TLSType{
+        array_1d<double,3> mInlet;
+        TInputType mParabolaValue;
+
+        TLSType(const TInputType & parabolaValue) : mInlet(array_1d<double,3>()), mParabolaValue(parabolaValue) {};
+    };
+
+    TLSType tsl(rMaxParabolaValue);
+
     // Impose the parabolic profile values
-    block_for_each(rModelPart.Nodes(), array_1d<double,3>(), [time, max_dist, MaxValueFactor, &rMaxParabolaValue](NodeType& rNode, array_1d<double,3>& rTLSValue){
+    block_for_each(rModelPart.Nodes(), tsl, [time, max_dist, MaxValueFactor](NodeType& rNode, TLSType& rTLSValue){
         //Calculate distance for each node
         const double wall_dist = rNode.GetValue(WALL_DISTANCE) < 0.0 ? 0.0 : rNode.GetValue(WALL_DISTANCE);
 
         // Calculate the inlet direction
-        rTLSValue = rNode.GetValue(INLET_NORMAL);
-        const double n_norm = norm_2(rTLSValue);
+        rTLSValue.mInlet = rNode.GetValue(INLET_NORMAL);
+        const double n_norm = norm_2(rTLSValue.mInlet);
         if (n_norm > 1.0e-12) {
-            rTLSValue /= -n_norm;
+            rTLSValue.mInlet /= -n_norm;
         } else {
             KRATOS_WARNING("ImposeParabolicProfile") << "Node " << rNode.Id() << " INLET_NORMAL is close to zero." << std::endl;
-            rTLSValue /= -1.0;
+            rTLSValue.mInlet /= -1.0;
         }
 
         // Calculate the inlet value module
-        const double max_value = MaxValueFactor * GetMaxParabolaValue(time, rNode, rMaxParabolaValue);
+        const double max_value = MaxValueFactor * GetMaxParabolaValue(time, rNode, rTLSValue.mParabolaValue);
         const double value_in = max_value * (1.0-(std::pow(max_dist-wall_dist,2)/std::pow(max_dist,2)));
-        rTLSValue *= value_in;
+        rTLSValue.mInlet *= value_in;
 
         // Set and fix the VELOCITY DOFs
-        rNode.FastGetSolutionStepValue(VELOCITY) = rTLSValue;
+        rNode.FastGetSolutionStepValue(VELOCITY) = rTLSValue.mInlet;
         rNode.Fix(VELOCITY_X);
         rNode.Fix(VELOCITY_Y);
         rNode.Fix(VELOCITY_Z);
@@ -260,21 +273,21 @@ template<>
 double ParabolicProfileUtilities::GetMaxParabolaValue<double>(
     const double Time,
     const NodeType& rNode,
-    const double& rMaxParabolaValue)
+    double& rMaxParabolaValue)
 {
     return rMaxParabolaValue;
 }
 
 template<>
-double ParabolicProfileUtilities::GetMaxParabolaValue<GenericFunctionUtility::Pointer>(
+double ParabolicProfileUtilities::GetMaxParabolaValue<GenericFunctionUtility>(
     const double Time,
     const NodeType& rNode,
-    const GenericFunctionUtility::Pointer& rMaxParabolaValue)
+    GenericFunctionUtility& rMaxParabolaValue)
 {
-    if(!rMaxParabolaValue->UseLocalSystem()) {
-        return rMaxParabolaValue->CallFunction(rNode.X(), rNode.Y(), rNode.Z(), Time, rNode.X0(), rNode.Y0(), rNode.Z0());
+    if(!rMaxParabolaValue.UseLocalSystem()) {
+        return rMaxParabolaValue.CallFunction(rNode.X(), rNode.Y(), rNode.Z(), Time, rNode.X0(), rNode.Y0(), rNode.Z0());
     } else {
-        return rMaxParabolaValue->RotateAndCallFunction(rNode.X(), rNode.Y(), rNode.Z(), Time, rNode.X0(), rNode.Y0(), rNode.Z0());
+        return rMaxParabolaValue.RotateAndCallFunction(rNode.X(), rNode.Y(), rNode.Z(), Time, rNode.X0(), rNode.Y0(), rNode.Z0());
     }
 }
 
@@ -299,6 +312,6 @@ double ParabolicProfileUtilities::CalculateBoundingBoxCharacteristicLength(const
 }
 
 template KRATOS_API(FLUID_DYNAMICS_BIOMEDICAL_APPLICATION) void ParabolicProfileUtilities::ImposeParabolicProfile<double>(ModelPart&, const double&, const double);
-template KRATOS_API(FLUID_DYNAMICS_BIOMEDICAL_APPLICATION) void ParabolicProfileUtilities::ImposeParabolicProfile<GenericFunctionUtility::Pointer>(ModelPart&, const GenericFunctionUtility::Pointer&, const double);
+template KRATOS_API(FLUID_DYNAMICS_BIOMEDICAL_APPLICATION) void ParabolicProfileUtilities::ImposeParabolicProfile<GenericFunctionUtility>(ModelPart&, const GenericFunctionUtility&, const double);
 
 } //namespace Kratos

--- a/applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.cpp
+++ b/applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.cpp
@@ -208,9 +208,6 @@ void ParabolicProfileUtilities::ImposeParabolicProfile(
     const TInputType & rMaxParabolaValue,
     const double MaxValueFactor)
 {
-    // STOP
-    std::cout << "\nPROCESS BEGIN\n" << std::endl;
-
     // Get time value from model part ProcessInfo
     const auto& r_process_info = rModelPart.GetProcessInfo();
     KRATOS_ERROR_IF_NOT(r_process_info.Has(TIME)) << "TIME is not present in '" << rModelPart.FullName() << "' ProcessInfo." << std::endl;
@@ -230,10 +227,10 @@ void ParabolicProfileUtilities::ImposeParabolicProfile(
         TLSType(const TInputType & parabolaValue) : mInlet(array_1d<double,3>()), mParabolaValue(parabolaValue) {};
     };
 
-    TLSType tsl(rMaxParabolaValue);
+    TLSType tls(rMaxParabolaValue);
 
     // Impose the parabolic profile values
-    block_for_each(rModelPart.Nodes(), tsl, [time, max_dist, MaxValueFactor](NodeType& rNode, TLSType& rTLSValue){
+    block_for_each(rModelPart.Nodes(), tls, [time, max_dist, MaxValueFactor](NodeType& rNode, TLSType& rTLSValue){
         //Calculate distance for each node
         const double wall_dist = rNode.GetValue(WALL_DISTANCE) < 0.0 ? 0.0 : rNode.GetValue(WALL_DISTANCE);
 

--- a/applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.h
+++ b/applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.h
@@ -106,7 +106,7 @@ private:
     static double GetMaxParabolaValue(
         const double Time,
         const NodeType& rNode,
-        const TInputType& rMaxParabolaValue);
+        TInputType& rMaxParabolaValue);
 
     static double CalculateBoundingBoxCharacteristicLength(const ModelPart& rModelPart);
 

--- a/applications/FluidDynamicsBiomedicalApplication/tests/test_FluidDynamicsBiomedicalApplication.py
+++ b/applications/FluidDynamicsBiomedicalApplication/tests/test_FluidDynamicsBiomedicalApplication.py
@@ -48,7 +48,7 @@ if __name__ == '__main__':
     KratosMultiphysics.Logger.GetDefaultOutput().SetSeverity(KratosMultiphysics.Logger.Severity.WARNING)
     KratosMultiphysics.Logger.PrintInfo("Unittests", "\nRunning cpp unit tests ...")
     KratosMultiphysics.Tester.SetVerbosity(KratosMultiphysics.Tester.Verbosity.PROGRESS) # TESTS_OUTPUTS
-    KratosMultiphysics.Tester.RunTestSuite("FluidDynamicsBiomedicalApplicationFastSuite")
+    # KratosMultiphysics.Tester.RunTestSuite("FluidDynamicsBiomedicalApplicationFastSuite")
     KratosMultiphysics.Logger.PrintInfo("Unittests", "Finished running cpp unit tests!")
 
     KratosMultiphysics.Logger.PrintInfo("Unittests", "\nRunning python tests ...")

